### DISCUSSION TITLE
how to set cert file to yarp for ssl

### DISCUSSION THREAD

**Discussion by dantemayson** [dotnet/yarp#2321] [General] [answered]
How can I implement ssl on yarp?
It is briefly explained in the yarp documentation, but I did not understand why I should do this

<html>
<body>
<h3 id="clientcert" style="box-sizing: border-box; font-family: &quot;Segoe UI&quot;, Tahoma, Helvetica, sans-serif; font-weight: 400; line-height: 1.1; color: rgb(51, 51, 51); margin-top: 25px; margin-bottom: 10px; font-size: 24px; overflow-wrap: break-word; word-break: break-word; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">ClientCert</h3><p style="box-sizing: border-box; margin: 0px 0px 10px; color: rgb(51, 51, 51); font-family: &quot;Segoe UI&quot;, Tahoma, Helvetica, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><strong style="box-sizing: border-box; font-weight: 700;">Forwards the client cert used on the inbound connection as a header to destination</strong></p><div class="table-responsive" style="box-sizing: border-box; min-height: 0.01%; overflow-x: auto; color: rgb(51, 51, 51); font-family: &quot;Segoe UI&quot;, Tahoma, Helvetica, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Key | Value | Required
-- | -- | --
ClientCert | The header name | yes

</div><p style="box-sizing: border-box; margin: 0px 0px 10px; color: rgb(51, 51, 51); font-family: &quot;Segoe UI&quot;, Tahoma, Helvetica, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Config:</p><pre style="box-sizing: border-box; font-size: 13px; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; overflow: auto; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(51, 51, 51); word-break: break-all; overflow-wrap: break-word; background-color: rgb(245, 245, 245); border: 0px; border-radius: 4px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><code class="lang-JSON hljs" style="box-sizing: border-box; font-size: inherit; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; padding: 0px; color: rgb(51, 51, 51); background-color: inherit; border-radius: 0px; white-space: pre-wrap; display: inline; background-image: initial; background-position: initial; background-size: initial; background-repeat: initial; background-attachment: initial; background-origin: initial; background-clip: initial; overflow-x: auto;">{ <span class="hljs-attr" style="box-sizing: border-box; color: rgb(121, 93, 163);">"ClientCert"</span>: <span class="hljs-string" style="box-sizing: border-box; color: rgb(214, 157, 133);">"X-Client-Cert"</span> }
</code></pre><p style="box-sizing: border-box; margin: 0px 0px 10px; color: rgb(51, 51, 51); font-family: &quot;Segoe UI&quot;, Tahoma, Helvetica, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Code:</p><pre style="box-sizing: border-box; font-size: 13px; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; overflow: auto; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(51, 51, 51); word-break: break-all; overflow-wrap: break-word; background-color: rgb(245, 245, 245); border: 0px; border-radius: 4px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><code class="lang-csharp hljs" style="box-sizing: border-box; font-size: inherit; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; padding: 0px; color: rgb(51, 51, 51); background-color: inherit; border-radius: 0px; white-space: pre-wrap; display: inline; background-image: initial; background-position: initial; background-size: initial; background-repeat: initial; background-attachment: initial; background-origin: initial; background-clip: initial; overflow-x: auto;">routeConfig = routeConfig.WithTransformClientCertHeader(headerName: <span class="hljs-string" style="box-sizing: border-box; color: rgb(214, 157, 133);">"X-Client-Cert"</span>);
</code></pre><pre style="box-sizing: border-box; font-size: 13px; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; overflow: auto; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(51, 51, 51); word-break: break-all; overflow-wrap: break-word; background-color: rgb(245, 245, 245); border: 0px; border-radius: 4px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><code class="lang-C# hljs" style="box-sizing: border-box; font-size: inherit; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; padding: 0px; color: rgb(51, 51, 51); background-color: inherit; border-radius: 0px; white-space: pre-wrap; display: inline; background-image: initial; background-position: initial; background-size: initial; background-repeat: initial; background-attachment: initial; background-origin: initial; background-clip: initial; overflow-x: auto;">transformBuilderContext.AddClientCertHeader(headerName: <span class="hljs-string" style="box-sizing: border-box; color: rgb(214, 157, 133);">"X-Client-Cert"</span>);
</code></pre><p style="box-sizing: border-box; margin: 0px 0px 10px; color: rgb(51, 51, 51); font-family: &quot;Segoe UI&quot;, Tahoma, Helvetica, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Example:</p><pre style="box-sizing: border-box; font-size: 13px; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; overflow: auto; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(51, 51, 51); word-break: break-all; overflow-wrap: break-word; background-color: rgb(245, 245, 245); border: 0px; border-radius: 4px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><code class="hljs arduino" style="box-sizing: border-box; font-size: inherit; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; padding: 0px; color: rgb(51, 51, 51); background-color: inherit; border-radius: 0px; white-space: pre-wrap; display: inline; background-image: initial; background-position: initial; background-size: initial; background-repeat: initial; background-attachment: initial; background-origin: initial; background-clip: initial; overflow-x: auto;">X-<span class="hljs-built_in" style="box-sizing: border-box;">Client</span>-Cert: SSdtIGEgY2VydGlmaWNhdGU...
</code></pre><p style="box-sizing: border-box; margin: 0px 0px 10px; color: rgb(51, 51, 51); font-family: &quot;Segoe UI&quot;, Tahoma, Helvetica, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">As the inbound and outbound connections are independent, there needs to be a way to pass any inbound client certificate to the destination server. This transform causes the client certificate taken from<span> </span><code style="box-sizing: border-box; font-size: 12.6px; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; padding: 2px 4px; color: rgb(113, 115, 116); background-color: rgb(241, 242, 243); border-radius: 4px;">HttpContext.Connection.ClientCertificate</code><span> </span>to be Base64 encoded and set as the value for the given header name. The destination server may need that certificate to authenticate the client. There is no standard that defines this header and implementations vary, check your destination server for support.</p><p style="box-sizing: border-box; margin: 0px 0px 10px; color: rgb(51, 51, 51); font-family: &quot;Segoe UI&quot;, Tahoma, Helvetica, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Servers do minimal validation on the incoming client certificate by default. The certificate should be validated either in the proxy or the destination, see the<span> </span><a href="https://docs.microsoft.com/aspnet/core/security/authentication/certauth" style="box-sizing: border-box; background-color: transparent; color: rgb(51, 122, 183); text-decoration: none; cursor: pointer;">client certificate auth</a><span> </span>docs for details.</p><p style="box-sizing: border-box; margin: 0px 0px 10px; color: rgb(51, 51, 51); font-family: &quot;Segoe UI&quot;, Tahoma, Helvetica, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">This transform will only apply if the client certificate is already present on the connection. See the<span> </span><a href="https://docs.microsoft.com/aspnet/core/security/authentication/certauth#optional-client-certificates" style="box-sizing: border-box; background-color: transparent; color: rgb(51, 122, 183); text-decoration: none; cursor: pointer;">optional certs doc</a><span> </span>if it needs to be requested from the client on a per-route basis.</p>
</body>
</html>

Where should I write these codes in the service?
Code:

```
routeConfig = routeConfig.WithTransformClientCertHeader(headerName: "X-Client-Cert");
transformBuilderContext.AddClientCertHeader(headerName: "X-Client-Cert");
```
Example:

`X-Client-Cert: SSdtIGEgY2VydGlmaWNhdGU...`

**Comment by Tratcher** [comment, marked as answer]
The `X-Client-Cert` description above is used to pass the client's client certificate through the proxy to the destination, not a certificate provided by YARP.

Do you mean that you want to accept inbound HTTPS/TLS connections and then proxy them? 

YARP doesn't configure the inbound connections, that's AspNetCore / Kestrel.
https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-7.0#listenoptionsusehttps


Or do you want YARP to authenticate to the destination server using a client certificate? If you want to add a certificate for YARP to use when authenticating to the destination then try this:
https://microsoft.github.io/reverse-proxy/articles/http-client-config.html#configuring-the-http-client
